google/datalab/bigquery/_table.py

```diff
@@ -595,20 +598,24 @@ def to_dataframe(self, start_row=0, max_rows=None):
     Returns:
       A Pandas dataframe containing the table data.
     """
-    fetcher = self._get_row_fetcher(start_row=start_row, max_rows=max_rows)
+    fetcher = self._get_row_fetcher(start_row=start_row,
+                                    max_rows=max_rows,
+                                    page_size=self._MAX_PAGE_SIZE)
     count = 0
     page_token = None
-    df = None
+
+    # Collect results of page fetcher in separate dataframe objects, then
+    # concatenate them to reduce the amount of copying
+    df_list = []
+
     while True:
       page_rows, page_token = fetcher(page_token, count)
       if len(page_rows):
         count += len(page_rows)
-        if df is None:
-          df = pandas.DataFrame.from_records(page_rows)
-        else:
-          df = df.append(page_rows, ignore_index=True)
+        df_list.append(pandas.DataFrame.from_records(page_rows))
```

Comment:
As one more potential speed comparison: did we consider just collating the list of rows (not slices of the final `DataFrame`), and then only creating a single `DataFrame` at the end? That is, something like

```python
rows = []
while True:
  ...
  if len(page_rows):
    rows.extend(page_rows)
  ...
df = pd.DataFrame.from_records(rows)
```